### ISSUE TITLE
POST /chats HTTP/1.1" 404 Not Found

### ISSUE THREAD

**Issue by peachkeel** [Quansight/ragna#371]
### Bug description

Through the UI, Ragna seems to work fine when creating a chat:

```
INFO:     127.0.0.1:39158 - "POST /chats HTTP/1.1" 200 OK
INFO:     127.0.0.1:39158 - "POST /chats/dd59fe2a-321c-484c-8329-41e9c6295a9e/prepare HTTP/1.1" 200 OK
INFO:     127.0.0.1:39158 - "GET /chats HTTP/1.1" 200 OK
```

However, when creating a chat via a command line script, there is a problem:

```
INFO:     127.0.0.1:49142 - "POST /document HTTP/1.1" 200 OK
INFO:     127.0.0.1:49142 - "PUT /document HTTP/1.1" 200 OK
INFO:     127.0.0.1:49142 - "POST /chats HTTP/1.1" 404 Not Found
```

### How to reproduce the bug?

This is the offending code in my script:

```
    chat = check(session.post(
        f"{RAGNA_URL}/chats",
        json={
            "name": CHAT_NAME,
            "documents": documents,
            "source_storage": CHAT_SOURCE_STORAGE,
            "assistant": CHAT_ASSISTANT,
            "params": {}
        }
    ))
```

And here is the exception it generates:

```
requests.exceptions.HTTPError: 404 Client Error: Not Found for url: http://xxxxx.xxx:31476/chats
```

[URL has been redacted]

### Versions and dependencies used

Ragna: 0.2.0
Python: 3.10.12
OS: 22.04.1-Ubuntu

**Comment by pmeier** [comment]
That is weird. I cannot reproduce. Here is the output of running `scripts/add_chats.py` (on the #372 branch) with a cold API start.

```
INFO:   RagnaDemoAuthentication: You can log in with any username and a matching password.
INFO:     Started server process [5589]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://127.0.0.1:31476 (Press CTRL+C to quit)
INFO:     127.0.0.1:54478 - "GET / HTTP/1.1" 200 OK
INFO:     127.0.0.1:54478 - "POST /token HTTP/1.1" 200 OK
INFO:     127.0.0.1:54478 - "POST /document HTTP/1.1" 200 OK
INFO:     127.0.0.1:54478 - "PUT /document HTTP/1.1" 200 OK
INFO:     127.0.0.1:54478 - "POST /document HTTP/1.1" 200 OK
INFO:     127.0.0.1:54478 - "PUT /document HTTP/1.1" 200 OK
INFO:     127.0.0.1:54478 - "POST /document HTTP/1.1" 200 OK
INFO:     127.0.0.1:54478 - "PUT /document HTTP/1.1" 200 OK
INFO:     127.0.0.1:54478 - "POST /document HTTP/1.1" 200 OK
INFO:     127.0.0.1:54478 - "PUT /document HTTP/1.1" 200 OK
INFO:     127.0.0.1:54478 - "POST /document HTTP/1.1" 200 OK
INFO:     127.0.0.1:54478 - "PUT /document HTTP/1.1" 200 OK
INFO:     127.0.0.1:54478 - "POST /chats HTTP/1.1" 200 OK
INFO:     127.0.0.1:54478 - "POST /chats/3fd0078f-c480-4c11-bde7-167fb3bbd076/prepare HTTP/1.1" 200 OK
INFO:     127.0.0.1:54478 - "POST /chats/3fd0078f-c480-4c11-bde7-167fb3bbd076/answer HTTP/1.1" 200 OK
INFO:     127.0.0.1:54478 - "POST /chats HTTP/1.1" 200 OK
INFO:     127.0.0.1:54478 - "POST /chats/8f296403-4eda-4908-af4a-b0d98de83b50/prepare HTTP/1.1" 200 OK
INFO:     127.0.0.1:54478 - "POST /chats/8f296403-4eda-4908-af4a-b0d98de83b50/answer HTTP/1.1" 200 OK
INFO:     127.0.0.1:54478 - "POST /chats/8f296403-4eda-4908-af4a-b0d98de83b50/answer HTTP/1.1" 200 OK
INFO:     127.0.0.1:54478 - "POST /chats/8f296403-4eda-4908-af4a-b0d98de83b50/answer HTTP/1.1" 200 OK
INFO:     127.0.0.1:54478 - "POST /chats HTTP/1.1" 200 OK
INFO:     127.0.0.1:54478 - "POST /chats/08397874-acc4-4d0e-bc9f-b92577b94b1f/prepare HTTP/1.1" 200 OK
INFO:     127.0.0.1:54478 - "POST /chats/08397874-acc4-4d0e-bc9f-b92577b94b1f/answer HTTP/1.1" 200 OK
INFO:     127.0.0.1:54478 - "POST /chats/08397874-acc4-4d0e-bc9f-b92577b94b1f/answer HTTP/1.1" 200 OK
INFO:     127.0.0.1:54478 - "GET /chats HTTP/1.1" 200 OK
```

Here is how we create the chat there, i.e. `POST /chats`

https://github.com/Quansight/ragna/blob/c6bdb0f9f5f71b37c0c312280839ab58d3059e34/scripts/add_chats.py#L51-L64

They look equivalent. The only thing that could produce a 404 here, should be a broken URL. Could you double check that the `RAGNA_URL` is set correctly and is the same as for the `POST /documents` and `PUT /documents` endpoints?

**Comment by peachkeel** [comment]
My mistake!

I was passing a non-existent assistant to Ragna via the script.

We might want to work on improving error messages/logging. The system is a little opaque when it comes to debugging.

**Comment by pmeier** [comment]
We are hitting this code block:

https://github.com/Quansight/ragna/blob/aa2a70263b373baaf40b4c41701c850e40873ad9/ragna/deploy/_api/core.py#L57-L65

I agree that 404 is kinda confusing. I mean it even through me off above. But the error message that we are sending with the status code makes it clear what happened

```
INFO:     127.0.0.1:51602 - "POST /document HTTP/1.1" 200 OK
INFO:     127.0.0.1:51602 - "PUT /document HTTP/1.1" 200 OK
INFO:     127.0.0.1:51602 - "POST /chats HTTP/1.1" 404 Not Found
```

```python
print(client.post("/chats", ...).json())
```

```
{'error': {'message': 'Unknown component, display_name=NonExistentAssistant'}}
```

From https://developer.mozilla.org/en-US/docs/Web/HTTP/Status#client_error_responses

> [404 Not Found](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/404)
>
> The server cannot find the requested resource. In the browser, this means the URL is not recognized. **In an API, this can also mean that the endpoint is valid but the resource itself does not exist.** (Emphasis mine)

So we are technically right here. But this means nothing if it taints the UX. What error code would you prefer?

**Comment by peachkeel** [comment]
I would say returning a `404` on a `GET` request makes sense. That is, an endpoint exists but the specific resource does not, where a _**resource**_ is defined as follows:

> The target of an HTTP request is called a "resource", whose nature isn't defined further; it can be a document, a photo, or anything else. Each resource is identified by a Uniform Resource Identifier (URI) used throughout HTTP for identifying resources.
>
>https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/Identifying_resources_on_the_Web

On a `POST` request, which is a processing request designed to produce a resource defined by a URI, I would say it makes a lot less sense (and is actually technically incorrect).

Better candidate status codes would be as follows:

- [`400 Bad Request`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/400)
- [`422 Unprocessable Content`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/422)
- [`500 Internal Server Error`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/500)

Of these, I think `422 Unprocessable Content` describes our error condition the best (i.e., proper syntax but bad semantics).

**Comment by pmeier** [comment]
SGTM. Do you want to send a PR?

**Comment by Tengal-Teemo** [comment]
@pmeier I'll take care of the PR

**Comment by Tengal-Teemo** [comment]
@pmeier @peachkeel #374

**Comment by pmeier** [comment]
Fixed in #374.
